R-package/inst/make-r-def.R

```diff
@@ -0,0 +1,98 @@
+# [description]
+#     Create a definition file (.def) from a .dll file, using objdump.
+#
+# [usage]
+#
+#     Rscript make-r-def.R something.dll something.def
+#
+# [references]
+#    * https://www.cs.colorado.edu/~main/cs1300/doc/mingwfaq.html
+
+args <- commandArgs(trailingOnly = TRUE)
+
+IN_DLL_FILE <- args[[1L]]
+OUT_DEF_FILE <- args[[2L]]
+DLL_BASE_NAME <- basename(IN_DLL_FILE)
+
+message(sprintf("Creating '%s' from '%s'", OUT_DEF_FILE, IN_DLL_FILE))
+
+# system() will not raise an R exception if the process called
+# fails. Wrapping it here to get that behavior.
+#
+# system() introduces a lot of overhead, at least on Windows,
+# so trying processx if it is available
+.pipe_shell_command_to_stdout <- function(command, args, out_file) {
+    has_processx <- suppressMessages({
+      suppressWarnings({
+        require("processx")  # nolint
+      })
+    })
+    if (has_processx) {
+        p <- processx::process$new(
+            command = command
+            , args = args
+            , stdout = out_file
+            , windows_verbatim_args = FALSE
+        )
+        invisible(p$wait())
+    } else {
+        message(paste0(
+          "Using system2() to run shell commands. Installing "
+          , "'processx' with install.packages('processx') might "
+          , "make this faster."
+        ))
+        exit_code <- system2(
+            command = command
+            , args = args
+            , stdout = out_file
+        )
+        if (exit_code != 0L) {
+            stop(paste0("Command failed with exit code: ", exit_code))
+        }
+    }
+    return(invisible(NULL))
+}
+
+# use objdump to dump all the symbols
+OBJDUMP_FILE <- "objdump-out.txt"
+.pipe_shell_command_to_stdout(
+    command = "objdump"
+    , args = c("-p", IN_DLL_FILE)
+    , out_file = OBJDUMP_FILE
+)
+
+objdump_results <- readLines(OBJDUMP_FILE)
+result <- file.remove(OBJDUMP_FILE)
```

Comment:
Seems that `result` is not used.

Reply:
this was just to suppress a log message (`TRUE` is printed when the file was removed). I'll just make it `invisible()` instead if it will reduce confusion

Reply:
fixed in https://github.com/microsoft/LightGBM/pull/3065/commits/709ddf3a0b0e5f909b6998d146ba2e0863d65102